File: src/main/scala/iog/psg/bcc/BccApiImpl.scala
package iog.psg.bcc

import java.time.ZonedDateTime

import akka.actor.ActorSystem
import akka.http.scaladsl.marshalling.{ Marshal, Marshaller }
import akka.http.scaladsl.model.HttpMethods._
import akka.http.scaladsl.model.Uri.Query
import akka.http.scaladsl.model._
import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport._
import io.circe.generic.auto._
import io.circe.generic.extras.Configuration
import iog.psg.bcc.BccApi.Order.Order

import scala.concurrent.{ ExecutionContext, Future }

private class BccApiImpl(baseUriWithPort: String)(implicit ec: ExecutionContext, as: ActorSystem)
    extends BccApi {

  import BccApiCodec._
  import BccApiCodec.ImplicitCodecs._
  import AddressFilter.AddressFilter
  import iog.psg.bcc.BccApi._

  private val addresses = s"${baseUriWithPort}addresses"
  private val proxy = s"${baseUriWithPort}proxy"
  private val wallets = s"${baseUriWithPort}wallets"
  private val network = s"${baseUriWithPort}network"
  private val stakePools = s"${baseUriWithPort}stake-pools"
  private def generateMigrationsUrl(walletId: String) = s"$wallets/$walletId/migrations"

  implicit val config: Configuration = Configuration.default.withSnakeCaseMemberNames

  /**
   * @inheritdoc
   */
  override def listWallets: BccApiRequest[Seq[Wallet]] = BccApiRequest(
    HttpRequest(
      uri = wallets,
      method = GET
    ),
    _.toWallets
  )

  /**
   * @inheritdoc
   */
  override def getWallet(walletId: String): BccApiRequest[Wallet] = BccApiRequest(
    HttpRequest(
      uri = s"$wallets/$walletId",
      method = GET
    ),
    _.toWallet
  )

  /**
   * @inheritdoc
   */
  override def updateName(walletId: String, name: String): Future[BccApiRequest[Wallet]] = {
    val body = Map("name" -> name)
    Marshal(body).to[RequestEntity].map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$wallets/$walletId",
          entity = marshalled,
          method = PUT
        ),
        _.toWallet
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def networkInfo: BccApiRequest[NetworkInfo] = BccApiRequest(
    HttpRequest(
      uri = s"${network}/information",
      method = GET
    ),
    _.toNetworkInfoResponse
  )

  /**
   * @inheritdoc
   */
  override def networkClock(forceNtpCheck: Option[Boolean] = None): BccApiRequest[NetworkClock] = {
    val url = s"$network/clock${forceNtpCheck.map(force => s"?forceNtpCheck=$force").getOrElse("")}"
    BccApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toNetworkClockResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def networkParameters(): BccApiRequest[NetworkParameters] =
    BccApiRequest(
      HttpRequest(
        uri = s"${network}/parameters",
        method = GET
      ),
      _.toNetworkParametersResponse
    )

  private def prepareCreateRestoreRequest[T](
    body: T
  )(implicit m: Marshaller[T, RequestEntity]) =
    Marshal(body).to[RequestEntity].map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$wallets",
          method = POST,
          entity = marshalled
        ),
        _.toWallet
      )
    }

  /**
   * @inheritdoc
   */
  override def createRestoreWallet(name: String,
                                   passphrase: String,
                                   mnemonicSentence: MnemonicSentence,
                                   mnemonicSecondFactor: Option[MnemonicSentence] = None,
                                   addressPoolGap: Option[Int] = None
  ): Future[BccApiRequest[Wallet]] =
    prepareCreateRestoreRequest(
      CreateRestore(
        name,
        passphrase,
        mnemonicSentence.mnemonicSentence,
        mnemonicSecondFactor.map(_.mnemonicSentence),
        addressPoolGap
      )
    )

  /**
   * @inheritdoc
   */
  override def createRestoreWalletWithKey(name: String,
                                          accountPublicKey: String,
                                          addressPoolGap: Option[Int] = None
  ): Future[BccApiRequest[Wallet]] =
    prepareCreateRestoreRequest(
      CreateRestoreWithKey(
        name,
        accountPublicKey,
        addressPoolGap
      )
    )

  /**
   * @inheritdoc
   */
  override def listAddresses(walletId: String,
                             state: Option[AddressFilter]
  ): BccApiRequest[Seq[WalletAddressId]] = {

    val baseUri = Uri(s"$wallets/${walletId}/addresses")

    val url = state.map { s =>
      baseUri.withQuery(Query("state" -> s.toString))
    }.getOrElse(baseUri)

    BccApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toWalletAddressIds
    )

  }

  /**
   * @inheritdoc
   */
  override def inspectAddress(addressId: String): BccApiRequest[WalletAddress] = {
    val url = Uri(s"$addresses/$addressId")

    BccApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toWalletAddress
    )
  }

  /**
   * @inheritdoc
   */
  override def listTransactions(walletId: String,
                                start: Option[ZonedDateTime] = None,
                                end: Option[ZonedDateTime] = None,
                                order: Order = Order.descendingOrder,
                                minWithdrawal: Option[Int] = None
  ): BccApiRequest[Seq[CreateTransactionResponse]] = {
    val baseUri = Uri(s"$wallets/${walletId}/transactions")

    val queries =
      Seq("start", "end", "order", "minWithdrawal").zip(Seq(start, end, order, minWithdrawal)).collect {
        case (queryParamName, order: Order)            => queryParamName -> order.toString
        case (queryParamName, Some(dt: ZonedDateTime)) => queryParamName -> zonedDateToString(dt)
        case (queryParamName, Some(minWith: Int))      => queryParamName -> minWith.toString
      }

    val uriWithQueries = baseUri.withQuery(Query(queries: _*))

    BccApiRequest(
      HttpRequest(
        uri = uriWithQueries,
        method = GET
      ),
      _.toCreateTransactionsResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def createTransaction(fromWalletId: String,
                                 passphrase: String,
                                 payments: Payments,
                                 metadata: Option[TxMetadataIn],
                                 withdrawal: Option[String]
  ): Future[BccApiRequest[CreateTransactionResponse]] = {

    val createTx = CreateTransaction(passphrase, payments.payments, metadata, withdrawal)

    Marshal(createTx).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$wallets/$fromWalletId/transactions",
          method = POST,
          entity = marshalled
        ),
        _.toCreateTransactionResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def estimateFee(fromWalletId: String,
                           payments: Payments,
                           withdrawal: Option[String],
                           metadataIn: Option[TxMetadataIn] = None
  ): Future[BccApiRequest[EstimateFeeResponse]] = {

    val estimateFees = EstimateFee(payments.payments, withdrawal, metadataIn)

    Marshal(estimateFees).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$wallets/$fromWalletId/payment-fees",
          method = POST,
          entity = marshalled
        ),
        _.toEstimateFeeResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def fundPayments(walletId: String,
                            payments: Payments
  ): Future[BccApiRequest[FundPaymentsResponse]] =
    Marshal(payments).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$wallets/${walletId}/coin-selections/random",
          method = POST,
          entity = marshalled
        ),
        _.toFundPaymentsResponse
      )
    }

  /**
   * @inheritdoc
   */
  override def getTransaction[T <: TxMetadataIn](walletId: String,
                                                 transactionId: String
  ): BccApiRequest[CreateTransactionResponse] = {

    val uri = Uri(s"$wallets/${walletId}/transactions/${transactionId}")

    BccApiRequest(
      HttpRequest(
        uri = uri,
        method = GET
      ),
      _.toCreateTransactionResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def deleteTransaction(walletId: String, transactionId: String): BccApiRequest[Unit] = {
    val uri = Uri(s"$wallets/${walletId}/transactions/${transactionId}")

    BccApiRequest(
      HttpRequest(
        uri = uri,
        method = DELETE
      ),
      _.toUnit
    )
  }

  /**
   * @inheritdoc
   */
  override def updatePassphrase(walletId: String,
                                oldPassphrase: String,
                                newPassphrase: String
  ): Future[BccApiRequest[Unit]] = {

    val uri = Uri(s"$wallets/${walletId}/passphrase")
    val updater = UpdatePassphrase(oldPassphrase, newPassphrase)

    Marshal(updater).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = uri,
          method = PUT,
          entity = marshalled
        ),
        _.toUnit
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def deleteWallet(
    walletId: String
  ): BccApiRequest[Unit] = {

    val uri = Uri(s"$wallets/${walletId}")

    BccApiRequest(
      HttpRequest(
        uri = uri,
        method = DELETE
      ),
      _.toUnit
    )
  }

  /**
   * @inheritdoc
   */
  override def getUTxOsStatistics(walletId: String): BccApiRequest[UTxOStatistics] = {
    val uri = Uri(s"$wallets/$walletId/statistics/utxos")
    BccApiRequest(
      HttpRequest(
        uri = uri,
        method = GET
      ),
      _.toUTxOStatisticsResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def postExternalTransaction(binary: String): BccApiRequest[PostExternalTransactionResponse] = {
    val uri = Uri(s"$proxy/transactions")
    BccApiRequest(
      HttpRequest(
        uri = uri,
        method = POST,
        entity = HttpEntity(binary).withContentType(ContentTypes.`application/octet-stream`)
      ),
      _.toPostExternalTransactionResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def migrateSophieWallet(walletId: String,
                                    passphrase: String,
                                    addresses: Seq[String]
  ): Future[BccApiRequest[Seq[MigrationResponse]]] = {
    val updater = SubmitMigration(passphrase = passphrase, addresses = addresses)
    Marshal(updater).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = generateMigrationsUrl(walletId),
          method = POST,
          entity = marshalled
        ),
        _.toSubmitMigrationsResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def getSophieWalletMigrationInfo(walletId: String): BccApiRequest[MigrationCostResponse] =
    BccApiRequest(
      HttpRequest(
        uri = generateMigrationsUrl(walletId),
        method = GET
      ),
      _.toMigrationCostResponse
    )

  /**
   * @inheritdoc
   */
  override def listStakePools(stake: Int): BccApiRequest[Seq[StakePool]] = {
    val url = Uri(s"$stakePools?stake=$stake")
    BccApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toStakePoolsResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def estimateFeeStakePool(walletId: String): BccApiRequest[EstimateFeeResponse] = {
    val url = Uri(s"$wallets/$walletId/delegation-fees")
    BccApiRequest(
      HttpRequest(
        uri = url,
        method = GET
      ),
      _.toEstimateFeeResponse
    )
  }

  /**
   * @inheritdoc
   */
  override def joinStakePool(walletId: String,
                             stakePoolId: String,
                             passphrase: String
  ): Future[BccApiRequest[MigrationResponse]] = {
    val updater = PassphraseRequest(passphrase = passphrase)
    Marshal(updater).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$stakePools/$stakePoolId/wallets/$walletId",
          method = PUT,
          entity = marshalled
        ),
        _.toSubmitMigrationResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def quitStakePool(walletId: String,
                             passphrase: String
  ): Future[BccApiRequest[MigrationResponse]] = {
    val updater = PassphraseRequest(passphrase = passphrase)
    Marshal(updater).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$stakePools/*/wallets/$walletId",
          method = DELETE,
          entity = marshalled
        ),
        _.toSubmitMigrationResponse
      )
    }
  }

  /**
   * @inheritdoc
   */
  override def getMaintenanceActions(): BccApiRequest[StakePoolMaintenanceActionsStatus] =
    BccApiRequest(
      HttpRequest(
        uri = s"$stakePools/maintenance-actions",
        method = GET
      ),
      _.toStakePoolMaintenanceActionsStatusResponse
    )

  /**
   * @inheritdoc
   */
  override def postMaintenanceAction(): Future[BccApiRequest[Unit]] = {
    val action = PostMaintenanceActionRequest("gc_stake_pools")
    Marshal(action).to[RequestEntity] map { marshalled =>
      BccApiRequest(
        HttpRequest(
          uri = s"$stakePools/maintenance-actions",
          method = POST,
          entity = marshalled
        ),
        _.toUnit
      )
    }
  }

}

File: src/main/java/iog/psg/bcc/jpi/BccApiImpl.java
package iog.psg.bcc.jpi;

import iog.psg.bcc.BccApiCodec;
import scala.Enumeration;
import scala.Some;
import scala.collection.immutable.IndexedSeq;
import scala.collection.immutable.Seq;
import scala.jdk.javaapi.CollectionConverters;

import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.CompletionStage;


public class BccApiImpl implements BccApi {

    private final iog.psg.bcc.BccApi api;
    private final HelpExecute helpExecute;

    private BccApiImpl() {
        helpExecute = null;
        api = null;
    }

    /**
     * BccApi constructor
     *
     * @param api iog.psg.bcc.BccApi instance
     * @param helpExecute og.psg.bcc.jpi.HelpExecute instance
     */
    public BccApiImpl(iog.psg.bcc.BccApi api, HelpExecute helpExecute) {
        this.helpExecute = helpExecute;
        this.api = api;
        Objects.requireNonNull(api, "Api cannot be null");
        Objects.requireNonNull(helpExecute, "HelpExecute cannot be null");
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.Wallet> createRestore(
            String name,
            String passphrase,
            List<String> mnemonicWordList,
            int addressPoolGap) throws BccApiException {
        return createRestore(name, passphrase, mnemonicWordList, null, addressPoolGap);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.Wallet> createRestore(
            String name,
            String passphrase,
            List<String> mnemonicWordList,
            List<String> mnemonicSecondFactor,
            int addressPoolGap) throws BccApiException {
        BccApiCodec.MnemonicSentence mnem = createMnemonic(mnemonicWordList);

        Optional<BccApiCodec.MnemonicSentence> mnemonicSecondaryFactorOpt = Optional.empty();
        if (mnemonicSecondFactor != null) {
            BccApiCodec.MnemonicSentence mnemonicSentence = createMnemonicSecondary(mnemonicSecondFactor);
            mnemonicSecondaryFactorOpt = Optional.of(mnemonicSentence);
        }

        return helpExecute.execute(
                api.createRestoreWallet(name, passphrase, mnem, option(mnemonicSecondaryFactorOpt), option(addressPoolGap))
        );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.Wallet> createRestoreWithKey(
            String name,
            String accountPublicKey,
            int addressPoolGap
    ) throws BccApiException {
        return helpExecute.execute(
                api.createRestoreWalletWithKey(name, accountPublicKey, option(addressPoolGap))
        );
    }

    /**
     * {@inheritDoc}
     */
   @Override
   public CompletionStage<BccApiCodec.CreateTransactionResponse> createTransaction(
           String fromWalletId,
           String passphrase,
           List<BccApiCodec.Payment> payments,
           BccApiCodec.TxMetadataIn metadata,
           String withdrawal
   ) throws BccApiException {

        return helpExecute.execute(api.createTransaction(fromWalletId, passphrase,
                new BccApiCodec.Payments(CollectionConverters.asScala(payments).toSeq()),
                option(metadata),
                option(withdrawal)));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.CreateTransactionResponse> createTransaction(
            String fromWalletId,
            String passphrase,
            List<BccApiCodec.Payment> payments
    ) throws BccApiException {

        return createTransaction(fromWalletId, passphrase, payments, null, "self");
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.Wallet> getWallet(
            String fromWalletId) throws BccApiException {

        return helpExecute.execute(
                api.getWallet(fromWalletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<Void> deleteWallet(
            String fromWalletId) throws BccApiException {

        return helpExecute.execute(
                api.deleteWallet(fromWalletId)).thenApply(x -> null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.CreateTransactionResponse> getTransaction(
            String walletId, String transactionId) throws BccApiException {

        return helpExecute.execute(
                api.getTransaction(walletId, transactionId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<Void> deleteTransaction(String walletId, String transactionId) throws BccApiException {
        return helpExecute.execute(api.deleteTransaction(walletId, transactionId)).thenApply(x -> null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.EstimateFeeResponse> estimateFee(
            String walletId, List<BccApiCodec.Payment> payments) throws BccApiException {
        return estimateFee(walletId, payments, "self", null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.EstimateFeeResponse> estimateFee(
            String walletId,
            List<BccApiCodec.Payment> payments,
            String withdrawal,
            BccApiCodec.TxMetadataIn metadata) throws BccApiException {

        return helpExecute.execute(
                api.estimateFee(walletId,
                        new BccApiCodec.Payments(CollectionConverters.asScala(payments).toSeq()),
                        option(withdrawal), option(metadata)));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.FundPaymentsResponse> fundPayments(
            String walletId, List<BccApiCodec.Payment> payments) throws BccApiException {
        return helpExecute.execute(
                api.fundPayments(walletId,
                        new BccApiCodec.Payments(CollectionConverters.asScala(payments).toSeq())));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<BccApiCodec.WalletAddressId>> listAddresses(
            String walletId, AddressFilter addressFilter) throws BccApiException {

        Optional<Enumeration.Value> addressFilterOpt = Optional.empty();
        if (addressFilter != null) {
            Enumeration.Value v = BccApiCodec.AddressFilter$.MODULE$.Value(addressFilter.name().toLowerCase());
            addressFilterOpt = Optional.of(v);
        }

        return helpExecute.execute(
                api.listAddresses(walletId, option(addressFilterOpt))).thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<BccApiCodec.WalletAddressId>> listAddresses(
            String walletId) throws BccApiException {
        return listAddresses(walletId, null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.WalletAddress> inspectAddress(
            String addressId) throws BccApiException {
        return helpExecute.execute(api.inspectAddress(addressId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<BccApiCodec.CreateTransactionResponse>> listTransactions(
            ListTransactionsParamBuilder builder) throws BccApiException {
        return helpExecute.execute(
                api.listTransactions(
                        builder.getWalletId(),
                        option(builder.getStartTime()),
                        option(builder.getEndTime()),
                        builder.getOrder(),
                        option(builder.getMinwithdrawal())))
                .thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<BccApiCodec.Wallet>> listWallets() throws BccApiException {
        return helpExecute.execute(
                api.listWallets())
                .thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<Void> updatePassphrase(
            String walletId,
            String oldPassphrase,
            String newPassphrase) throws BccApiException {

        return helpExecute.execute(api.updatePassphrase(walletId, oldPassphrase, newPassphrase)).thenApply(x -> null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.Wallet> updateName(
            String walletId,
            String name) throws BccApiException {
        return helpExecute.execute(api.updateName(walletId, name));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.NetworkInfo> networkInfo() throws BccApiException {
        return helpExecute.execute(api.networkInfo());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.NetworkClock> networkClock() throws BccApiException {
        return networkClock(null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.NetworkClock> networkClock(Boolean forceNtpCheck) throws BccApiException {
        return helpExecute.execute(api.networkClock(option(forceNtpCheck)));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.NetworkParameters> networkParameters() throws BccApiException {
        return helpExecute.execute(api.networkParameters());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.UTxOStatistics> getUTxOsStatistics(String walletId) throws BccApiException {
        return helpExecute.execute(api.getUTxOsStatistics(walletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.PostExternalTransactionResponse> postExternalTransaction(String binary) throws BccApiException {
        return helpExecute.execute(api.postExternalTransaction(binary));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<BccApiCodec.MigrationResponse>> migrateSophieWallet(String walletId, String passphrase, List<String> addresses) throws BccApiException {
        IndexedSeq<String> addressesList = CollectionConverters.asScala(addresses).toIndexedSeq();
        CompletionStage<Seq<BccApiCodec.MigrationResponse>> response = helpExecute.execute(api.migrateSophieWallet(walletId, passphrase, addressesList));
        return response.thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.MigrationCostResponse> getSophieWalletMigrationInfo(String walletId) throws BccApiException {
        return helpExecute.execute(api.getSophieWalletMigrationInfo(walletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<List<BccApiCodec.StakePool>> listStakePools(Integer stake) throws BccApiException {
        CompletionStage<Seq<BccApiCodec.StakePool>> stakePools = helpExecute.execute(api.listStakePools(stake));
        return stakePools.thenApply(CollectionConverters::asJava);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.EstimateFeeResponse> estimateFeeStakePool(String walletId) throws BccApiException {
        return helpExecute.execute(api.estimateFeeStakePool(walletId));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.MigrationResponse> joinStakePool(String walletId, String stakePoolId, String passphrase) throws BccApiException {
        return helpExecute.execute(api.joinStakePool(walletId, stakePoolId, passphrase));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.MigrationResponse> quitStakePool(String walletId, String passphrase) throws BccApiException {
        return helpExecute.execute(api.quitStakePool(walletId, passphrase));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public CompletionStage<BccApiCodec.StakePoolMaintenanceActionsStatus> getMaintenanceActions() throws BccApiException {
        return helpExecute.execute(api.getMaintenanceActions());
    }

    @Override
    public CompletionStage<Void> postMaintenanceAction() throws BccApiException {
        return helpExecute.execute(api.postMaintenanceAction()).thenApply(x -> null);
    }

    private static <T> scala.Option<T> option(final T value) {
        return (value != null) ? new Some<T>(value) : scala.Option.apply((T) null);
    }

    private static <T> scala.Option<T> option(final Optional<T> value) {
        return value.map(BccApiImpl::option).orElse(scala.Option.apply((T) null));
    }

    private static BccApiCodec.GenericMnemonicSentence createMnemonic(List<String> wordList) {
        return new BccApiCodec.GenericMnemonicSentence(
                CollectionConverters.asScala(wordList).toIndexedSeq()
        );
    }

    private static BccApiCodec.GenericMnemonicSecondaryFactor createMnemonicSecondary(List<String> wordList) {
        return new BccApiCodec.GenericMnemonicSecondaryFactor(
                CollectionConverters.asScala(wordList).toIndexedSeq()
        );
    }

}
